### ISSUE TITLE
Launching a runtime workbench fails with UnsatisfiedLinkError: Could not load SWT library

### ISSUE THREAD

**Issue by betamaxbandit** [eclipse-platform/eclipse.platform.swt#1004]
@HannesWell
cc: @jonahgraham 

After installing Eclipse using the oomph installer, and attempting to launch a runtime workbench, the launch fails and the following message[1] is reported:

  java.lang.UnsatisfiedLinkError: Could not load SWT library. Reasons: 
    C:\Users\betamax\platform_jdt_committers\git\eclipse.platform.swt\binaries\org.eclipse.swt.win32.win32.x86_64\swt-win32-4964r6.dll: %1 is not a valid Win32 application

I inspected the contents of C:\Users\betamax\.swt\lib\win32\x86_64\swt-win32-4964r6.dll and found it wasn't an actual DLL. It contains the following text.

  version https://git-lfs.github.com/spec/v1
  oid sha256:e9107ea8fecca50d2feccbfb00bc49964c211f0d1f221b73d6e94bcb92534b35
  size 771464

I read the post here:
[platform-dev] SWT native binaries merged into main SWT repository using Git-LFS
https://www.eclipse.org/lists/platform-dev/msg03951.html

Things I've tried which have not fixed the issue:

"Enable LFS support globally" : https://github.com/eclipse-egit/egit/wiki/User-Guide#git-lfs-support

Restarted eclipse.

Help > Perform setup tasks.

Installed and updated lfs using CLI git terminal:
  $ ~/platform_jdt_committers/git/eclipse.platform.swt (master)
  $ git lfs update
  Updated Git hooks.

Env:
  Oomph Installer version: 1.32.0 Build 388
  Installing eclipse for committers 2023-12
  Product version: Latest (4.31 - 2024-03)
  Project: Platform
  Java: jdk 17.10
  Windows 11

**Workaround (or) Additional context**
None; blocked by this.

[1] Eclipse console output
WARNING: Using incubator modules: jdk.incubator.vector, jdk.incubator.foreign
SLF4J(W): No SLF4J providers were found.
SLF4J(W): Defaulting to no-operation (NOP) logger implementation
SLF4J(W): See https://www.slf4j.org/codes.html#noProviders for further details.
!SESSION 2024-01-24 17:13:04.085 -----------------------------------------------
eclipse.buildId=unknown
java.version=17.0.5
java.vendor=Oracle Corporation
BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=en_GB
Framework arguments:  -product org.eclipse.sdk.ide
Command-line arguments:  -product org.eclipse.sdk.ide -data C:\Users\betamax\platform_jdt_committers\wsPlatform2024-03b/../runtime-workspace -dev file:C:/Users/betamax/platform_jdt_committers/wsPlatform2024-03b/.metadata/.plugins/org.eclipse.pde.core/Runtime Workspace/dev.properties -os win32 -ws win32 -arch x86_64 -consoleLog

!ENTRY org.eclipse.osgi 4 0 2024-01-24 17:13:07.183
!MESSAGE Application error
!STACK 1
java.lang.UnsatisfiedLinkError: Could not load SWT library. Reasons: 
	C:\Users\betamax\platform_jdt_committers\git\eclipse.platform.swt\binaries\org.eclipse.swt.win32.win32.x86_64\swt-win32-4964r6.dll: %1 is not a valid Win32 application
	no swt-win32 in java.library.path: C:\Program Files\Java\jdk-17.0.5\bin;C:\WINDOWS\Sun\Java\bin;C:\WINDOWS\system32;C:\WINDOWS;C:/Program Files/Java/jdk-17.0.5/bin/server;C:/Program Files/Java/jdk-17.0.5/bin;C:\Program Files\Common Files\Oracle\Java\javapath;C:\Program Files\Eclipse Adoptium\jdk-11.0.15.10-hotspot\bin;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Users\betamax\AppData\Local\Microsoft\WindowsApps;C:\Users\betamax\AppData\Local\Programs\Git\cmd;C:\Users\betamax\AppData\Local\Programs\Microsoft VS Code\bin;;C:\Users\betamax\eclipse-installer4;;.
	no swt in java.library.path: C:\Program Files\Java\jdk-17.0.5\bin;C:\WINDOWS\Sun\Java\bin;C:\WINDOWS\system32;C:\WINDOWS;C:/Program Files/Java/jdk-17.0.5/bin/server;C:/Program Files/Java/jdk-17.0.5/bin;C:\Program Files\Common Files\Oracle\Java\javapath;C:\Program Files\Eclipse Adoptium\jdk-11.0.15.10-hotspot\bin;C:\Program Files (x86)\Common Files\Oracle\Java\javapath;C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Users\betamax\AppData\Local\Microsoft\WindowsApps;C:\Users\betamax\AppData\Local\Programs\Git\cmd;C:\Users\betamax\AppData\Local\Programs\Microsoft VS Code\bin;;C:\Users\betamax\eclipse-installer4;;.
	Can't load library: C:\Users\betamax\.swt\lib\win32\x86_64\swt-win32-4964r6.dll
	Can't load library: C:\Users\betamax\.swt\lib\win32\x86_64\swt-win32.dll
	Can't load library: C:\Users\betamax\.swt\lib\win32\x86_64\swt.dll
	C:\Users\betamax\.swt\lib\win32\x86_64\swt-win32-4964r6.dll: %1 is not a valid Win32 application

	at org.eclipse.swt.internal.Library.loadLibrary(Library.java:345)
	at org.eclipse.swt.internal.Library.loadLibrary(Library.java:254)
	at org.eclipse.swt.internal.C.<clinit>(C.java:19)
	at org.eclipse.swt.internal.win32.STARTUPINFO.<clinit>(STARTUPINFO.java:42)
	at org.eclipse.swt.widgets.Display.<clinit>(Display.java:148)
	at org.eclipse.ui.internal.Workbench.createDisplay(Workbench.java:730)
	at org.eclipse.ui.PlatformUI.createDisplay(PlatformUI.java:185)
	at org.eclipse.ui.internal.ide.application.IDEApplication.createDisplay(IDEApplication.java:182)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:125)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:208)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:143)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:109)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:439)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:271)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:651)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:588)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1459)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1432)
An error has occurred. See the log file
C:\Users\betamax\platform_jdt_committers\runtime-workspace\.metadata\.log.

**Comment by betamaxbandit** [comment]
Hi @HannesWell ,

In your https://www.eclipse.org/lists/platform-dev/msg03951.html you say this:

>If you are using SWT's Oomph setup (and EGit), just run it again and after a restart you should be ready to go.

What do you mean by just run it again? Run what?

Cheers John

**Comment by iloveeclipse** [comment]
> After installing Eclipse using the oomph installer

How did you get your SWT git repo? It should be "git lfs" enabled.

**Comment by betamaxbandit** [comment]
> > After installing Eclipse using the oomph installer
> 
> How did you get your SWT git repo? It should be "git lfs" enabled.

I chose Platform in the oomph projects page, which includes the SWT git repo. Then I just let it install.
The [platform-dev] post suggests this should just work out-of-the-box.

**Comment by merks** [comment]
I tried this:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/fc4032df-6582-402d-9bd1-6f6b96f57f8d)

and this:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/9e078e32-fbd9-46af-857d-474edc39bf1a)

And that works for me:

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/c011df1c-69f1-4963-8301-d40f4ba1a593)

**Comment by jonahgraham** [comment]
It just worked for me too - not sure what starting point your computer had that meant the lfs files did not get fetched.

I see you did `git lfs install`. Did you run oomph installer after doing that? 
What is the contents of your `~/.gitconfig` file?

**Comment by betamaxbandit** [comment]
Thanks guys for trying this.
So I just tried @merks 's SWT only oomph install and that worked; I'm able to launch o.e.platform.ide runtime workbench.
However, I notice that during launch it doesn't create .swt\lib\win32\x86_64\swt-win32-4964r6.dll at all for me, so I'm not sure whether this is a fair test of the problem!

> What is the contents of your ~/.gitconfig file?
'''
[filter "lfs"]
useJGitBuiltin = true
required = true
'''

I'm using a clean machine (Windows 11), which has no other dev tools (eg: CLI git) installed.

Just checking, but is the oomph installer expected to work on a clean machine? Are there any prerequisites apart from Java?

**Comment by merks** [comment]
The installer has a built-in jre and one can choose a justj jre for the installation. But the jre task requires a jre/jdk on the machine.

**Comment by HannesWell** [comment]
> However, I notice that during launch it doesn't create .swt\lib\win32\x86_64\swt-win32-4964r6.dll at all for me, so I'm not sure whether this is a fair test of the problem!
> 
This is probably not required when launching SWT form a dev eclipse. This is probably only used when the native binaries are within a jar, since they can only be loaded as plain files.
Side-note: we could consider to install the fragments with Eclipse-BundleShape: dir to have P2 unzip the jars during installation.

> > What is the contents of your ~/.gitconfig file?
> '''
> [filter "lfs"]
> useJGitBuiltin = true
> required = true
> '''

The entry `useJGitBuiltin = true`tells me that git.lfs is properly installed and activated.

Can you check the content of some of the native binaries (.dll, .so, ...) if the pointer file was  replaced, i.e. it's content is something binary and not a pain text file with only an object ID (oid).

Given that actually everything is set up, maybe checking out another branch or commit and the then again the master retriggers the pointer-file replacement.

> 
> Just checking, but is the oomph installer expected to work on a clean machine? Are there any prerequisites apart from Java?
> 
For Java development in SWT there should be no requirements if you use Oomph. But I can check it again with a clean user home and fresh Oomph provised eclipse. Or @merks did you do it already in a completely clean environment?

**Comment by merks** [comment]
I didn’t test in a 100% clean home. But pretty close to that state.

**Comment by betamaxbandit** [comment]
> It just worked for me too - not sure what starting point your computer had that meant the lfs files did not get fetched.

@jonahgraham , did you do the test that @merks did (just the SWT component) or the entire Platform project with child components like in my original post?

**Comment by merks** [comment]
FYI, I did test the entire platform from scratch just last week (January 20th) and that works fine too...

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/a2132145-6f7b-4222-b7d8-dd0a16c7159b)

![image](https://github.com/eclipse-platform/eclipse.platform.swt/assets/208716/b3ac1f45-af12-4760-978f-117866c6b10e)

**Comment by betamaxbandit** [comment]
I have retested this on my clean machine and it's all working now.
This was due to my misunderstanding how the [filter "lfs"] param in the ~/.gitconfig file operated. I did not realise I should re-install using oomph from scratch once this setting had been written to the .gitconfig file.
Apologies for the noise.

**Comment by merks** [comment]
FYI, probably you only needed to run Help -> Perform Setup Tasks...
